### ISSUE TITLE
sudo should ask for password at beginning of 'vagrant up', not in the middle.

### ISSUE THREAD

**Issue by blobhouse** [cogitatio/vagrant-hostsupdater#55]
Hi,

We've started using this plugin with Vagrant under OS X at our office, and it works well.  But one issue that keeps coming up is that some time after running "vagrant up", 'sudo' requests the user's password (in order to edit /etc/hosts) in the middle of our fairly long provisioning process. The length of that process wouldn't be an issue if a user could start it and walk away, but instead they have to wait around an indeterminate period of time before they're asked to enter their password.

I solve this by running "sudo echo foo; vagrant up" so that sudo caches my password.  It would be much better if the plugin did that internally: ask for the user's password as early in the process as possible, and let sudo cache it until it's needed.

**Comment by Johnstonian** [comment]
This would most likely have to happen at the Vagrant level. Not sure a plugin can make a sudo password request at the beginning of "vagrant up". This is a good idea though. Happens to me all the time and its frustrating. Maybe open an issue with the Vagrant project? https://github.com/mitchellh/vagrant

**Comment by dasrecht** [comment]
Hi there

well we had the same problem and circumvented this the same way it's done when you use NFS mounting on vagrant itself (https://github.com/mitchellh/vagrant/blob/681ba1c3d754b5ec9793a8811ab40eba95c11668/website/docs/source/v2/synced-folders/nfs.html.md)

adding this to the sudoers file should work fine
```
# Allow passwordless startup of Vagrant with vagrant-hostsupdater.
Cmnd_Alias VAGRANT_HOSTS_ADD = /bin/sh -c echo "*" >> /etc/hosts
Cmnd_Alias VAGRANT_HOSTS_REMOVE = /usr/bin/sed -i -e /*/ d /etc/hosts
%admin ALL=(root) NOPASSWD: VAGRANT_HOSTS_ADD, VAGRANT_HOSTS_REMOVE
```

Probably we should document this into the readme file and update it when the hostsupdater commands change.

What do you think?

**Comment by meitar** [comment]
> Probably we should document this into the readme file and update it when the hostsupdater commands change.
> 
> What do you think?

I just ran into this same issue and independently used the same pattern in my `sudoers` file to make this fix. My setup is slightly different, though, in that on my dev machine, my user isn't in the `admin` group. Seems to me like it would make sense to encourage users who aren't already running as part of an `admin` group for their day-to-day development environment to continue that security best practice.

#73 is a pull request that prints a generalized version of the above advice into `STDERR` when this issue is encountered.

**Comment by cgsmith** [comment]
Going to close this for now.  The main issue was addressed in version [1.0.0](../releases/tag/1.0.0).  Host entries are added before any provisioner runs so the user is prompted fairly quickly to enter the sudo password.

**Comment by aijanai** [comment]
Could you please specify that it must be workstation's sudo password?
I was stuck in thinking that it was asking for sudo permission inside vagrant to modify its hosts file and not the one of my host:

```
==> machine: [vagrant-hostsupdater] This operation requires administrative access. You may skip it by manually adding equivalent entries to the hosts file.
Password:
```
 Maybe we could change the message to something more clarifying like: `This operation requires administrative access on your vagrant workstation`.

Thanks
